feat: Allow specific indexers in EntityIndexerRegistry

NielDuysters opened the pull request:
### 1. Why is this change necessary?
When you want to exclude indexers to execute on the update of an Entity, you currently have to exclude all the indexers manually.
```php
$context->addExtension(EntityIndexerRegistry::EXTENSION_INDEXER_SKIP,
    new ArrayEntity(
        [
            // All except STOCK_UPDATER.
            ProductIndexer::INHERITANCE_UPDATER,
            ProductIndexer::VARIANT_LISTING_UPDATER,
            ProductIndexer::CHILD_COUNT_UPDATER,
            ProductIndexer::MANY_TO_MANY_ID_FIELD_UPDATER,
            ProductIndexer::CATEGORY_DENORMALIZER_UPDATER,
            ProductIndexer::CHEAPEST_PRICE_UPDATER,
            ProductIndexer::RATING_AVERAGE_UPDATER,
            ProductIndexer::STREAM_UPDATER,
            ProductIndexer::SEARCH_KEYWORD_UPDATER,
        ]
    ),
);

$this->productRepository->update([
    [
        'id' => $productId,
        'stock' => $newStock,
    ]
], $context);
```

### 2. What does this change do, exactly?
This change allows you to whitelist only the indexers you want to execute.
```php
$context->addExtension(EntityIndexerRegistry::EXTENSION_INDEXER_ONLY,
    new ArrayEntity(
        [
            // Only execute STOCK_UPDATER.
            ProductIndexer::STOCK_UPDATER
        ]
    ),
);
        
$this->productRepository->update([
    [
        'id' => $productId,
        'stock' => $newStock,
    ]
], $context);
```

### 3. Describe each step to reproduce the issue or behaviour.
I made a simple subscriber to test this code. Add an item to the cart to trigger.
<details>
<summary>View code</summary>

```php
<?php declare(strict_types=1);

namespace DemoWebshop\Subscriber;

use Psr\Log\LoggerInterface;
use Shopware\Core\Checkout\Cart\Event\AfterLineItemAddedEvent;
use Shopware\Core\Content\Product\DataAbstractionLayer\ProductIndexer;
use Shopware\Core\Framework\DataAbstractionLayer\EntityRepository;
use Shopware\Core\Framework\DataAbstractionLayer\Indexing\EntityIndexerRegistry;
use Shopware\Core\Framework\DataAbstractionLayer\Search\Criteria;
use Shopware\Core\Framework\Struct\ArrayStruct;
use Symfony\Component\EventDispatcher\EventSubscriberInterface;

class MySubscriber implements EventSubscriberInterface
{
    public function __construct(
        private readonly EntityRepository $productRepository,
        private readonly LoggerInterface $logger
    )
    {
    }

    public static function getSubscribedEvents(): array
    {
        // Return the events to listen to as array like this:  <event to listen to> => <method to execute>
        return [
            AfterLineItemAddedEvent::class => 'onAfterLineItemAdded',
        ];
    }

    public function onAfterLineItemAdded(AfterLineItemAddedEvent $event): void
    {
        $context = $event->getContext();
        $cart = $event->getCart();
        $newStock = 30;

        $context->addExtension(EntityIndexerRegistry::EXTENSION_INDEXER_SKIP, // This PR allows the usage of EXTENSION_INDEXER_ONLY
            new ArrayStruct(
                [
                    ProductIndexer::STOCK_UPDATER
                ]
            ),
        );

        foreach ($cart->getLineItems() as $cartItem) {
            try {
                /** @var Product $product */
                $product = $this->productRepository->search(
                    (new Criteria([$cartItem->getId()]))->addAssociation('stock'),
                    $context
                )->getEntities()->first();

                $this->productRepository->update([
                    [
                        'id' => $cartItem->getId(),
                        'stock' => $newStock,
                    ]
                ], $context);
            } catch (\Exception $exception) {
                $this->logger->error($exception->getMessage());
            }
        }
    }
}
```
</details>

### 4. Please link to the relevant issues (if any).
N/A

### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](changelog/_unreleased/2025-04-12-entityindexerregistry-extension-indexer-only.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfill them.

NielDuysters committed:
entityindexerregistry: Add EXTENSION_INDEXER_ONLY

Allow to index specific indexes instead of blacklisting the ones needed to be excluded.

NielDuysters committed:
entityindexerregistry: Update test

NielDuysters committed:
entityindexerregistry: Changelog

NielDuysters committed:
entityindexerregistry: Small refactor - Remove obsolete if statements

These if-statements are obsolete because the checks for the extension is done in the body of the called methods as well.

We can consider reverting this commit and keep the if-statements for readability.

NielDuysters committed:
entityindexerregistry: Small refactor - Variable naming convention

NielDuysters committed:
entityindexerregistry: Small refactor - Doc-block type annotation

mitelg commented on changelog/_unreleased/2025-04-12-entityindexerregistry-extension-indexer-only.md:
```diff
@@ -0,0 +1,9 @@
+---
+title: Add EntityIndexerRegistry::EXTENSION_INDEXER_ONLY
+issue: N/A
```
```suggestion
```

mitelg commented on changelog/_unreleased/2025-04-12-entityindexerregistry-extension-indexer-only.md:
```diff
@@ -0,0 +1,9 @@
+---
+title: Add EntityIndexerRegistry::EXTENSION_INDEXER_ONLY
```
```suggestion
title: Add possibility to the context to define which specific indexer should run
```

mitelg commented on changelog/_unreleased/2025-04-12-entityindexerregistry-extension-indexer-only.md:
```diff
@@ -0,0 +1,9 @@
+author_github: @nielduysters
+---
+# Core
+* Added `EntityIndexerRegistry::EXTENSION_INDEXER_ONLY` to the `EntityIndexerRegistry` to allow specific indexes.
```
```suggestion
* Added `EntityIndexerRegistry::EXTENSION_INDEXER_ONLY` to the `EntityIndexerRegistry` to allow specific indexers.

# Upgrade Information

## Add possibility to specify indexer in tontext
< Add some information and how to use it here>
```

mitelg commented on src/Core/Framework/DataAbstractionLayer/Indexing/EntityIndexerRegistry.php:
```diff
@@ -145,7 +146,9 @@ public function refresh(EntityWrittenContainerEvent $event): void
             $message->isFullIndexing = false;
+
             self::addSkips($message, $context);
+            self::addOnlies($message, $indexer->getOptions(), $context);
```
if we apply the the only allowed indexers after the skipped ones, it is a different behaviour than in the `index` method, where we already consider specified indexer. 

there the skipped ones are first checked and then the allowed ones. 
so here we need to apply first the allowed ones, and then check if some of them are skipped. 

this should also be considered in a test

NielDuysters replied:
**Commit:** https://github.com/shopware/shopware/pull/8929/commits/edfda09b0734fb39b4f2307550c782436fcc1a2b

I updated the code. Now we first execute `addOnlies` which calls `$message->setSkip` to fully overwrite the `$skips` property in `$message`. IG this full overwrite is necessary because AFAICT Shopware configures some indexers to be skipped by default?

After `addOnlies` we execute `addSkips` which calls `$message->addSkip()` to add all the configured indexers to be skipped. Meaning that if we have `indexer1` configured as 'only', but `indexer1` is also configured as 'skip', it will be skipped.

I updated the `testRefreshMethod` to check that both `setSkip` and `addSkip` is called with the correct indexers on `$messageMock`.

mitelg reviewed (changes requested):
thanks for your contribution @NielDuysters 

I left some comments: 

the new flag should also be considered in the sync API: https://github.com/shopware/shopware/blob/trunk/src/Core/Framework/Api/Sync/SyncService.php#L39

NielDuysters committed:
entityindexerregistry: Update changelog/_unreleased/2025-04-12-entityindexerregistry-extension-indexer-only.md

Suggested change from PR reviewer.

Co-authored-by: Michael Telgmann <mitelg@users.noreply.github.com>

NielDuysters committed:
entityindexerregistry: Update changelog/_unreleased/2025-04-12-entityindexerregistry-extension-indexer-only.md

Suggested change by PR reviewer.

Co-authored-by: Michael Telgmann <mitelg@users.noreply.github.com>

NielDuysters committed:
entityindexerregistry: Update changelog/_unreleased/2025-04-12-entityindexerregistry-extension-indexer-only.md

Suggested change by PR reviewer.

Co-authored-by: Michael Telgmann <mitelg@users.noreply.github.com>

NielDuysters committed:
entityindexerregistry: Check skips after onlies

Previously the `addOnlies` fully overwrote the indexers configured to skip using `addSkips`.

We now first apply the onlies and then check what indexers needs to be skipped as well.

NielDuysters commented:
@mitelg 

Updated PR. Regarding the Sync API. IG I'd have to update the constructor of `SyncBehavior`? Will do this later, seems like this can cause some code regression.

Also, why, in `EntityIndexerRegistry`, do we prefer `ArrayEntity` over `ArrayStruct`?

With `ArrayEntity` the usage would look like:
```
        $context->addExtension(EntityIndexerRegistry::EXTENSION_INDEXER_ONLY,
            new ArrayEntity(['onlies' => [
                    ProductIndexer::STOCK_UPDATER,
                ]
            ]),
        );
```

While with `ArrayStruct` it could be simplified too:
```
        $context->addExtension(EntityIndexerRegistry::EXTENSION_INDEXER_ONLY,
            new ArrayStruct([
                ProductIndexer::STOCK_UPDATER,
            ]),
        );
```

Maybe we should allow both?

mitelg commented:
> Also, why, in `EntityIndexerRegistry`, do we prefer `ArrayEntity` over `ArrayStruct`?

hey @NielDuysters 
thanks for the update, I will take a look now. Regarding this question, I have no clue, tbh. To make it not confusing, I would just use the same mechanics as the "skip" stuff. We can improve here later in a second step and new PR.

mitelg commented on changelog/_unreleased/2025-04-12-entityindexerregistry-extension-indexer-only.md:
```diff
@@ -0,0 +1,13 @@
+# Upgrade Information
+
+## Add possibility to specify indexer in tontext
+< Add some information and how to use it here>
```
of course you need to fill that with life :grimacing: that was just a placeholder from my side

mitelg commented on changelog/_unreleased/2025-04-12-entityindexerregistry-extension-indexer-only.md:
```diff
@@ -0,0 +1,13 @@
+
+# Upgrade Information
+
+## Add possibility to specify indexer in tontext
```
```suggestion
## Add possibility to specify indexer in context
```

mitelg committed:
Update changelog/_unreleased/2025-04-12-entityindexerregistry-extension-indexer-only.md

mitelg commented:
hey @NielDuysters 

any update on this? :slightly_smiling_face:

mitelg committed:
Merge branch 'trunk' into feat/entityindexerregistry-extension-indexer-only

BrocksiNet commented on src/Core/Framework/DataAbstractionLayer/Indexing/EntityIndexerRegistry.php:
```diff
@@ -163,7 +166,29 @@ public static function addSkips(EntityIndexingMessage $message, Context $context
+    /**
+     * @param array<string> $options
+     */
+    public static function addOnlies(EntityIndexingMessage $message, array $options, Context $context): void
```
In the PR, we are talking about "Allow specific indexers" and now we call the function `addOnlies`?

I would suggest renaming it to `addOnlyAllowedIndices`.

NielDuysters replied:
@BrocksiNet ✅  FTFY. Thx!

mitelg commented:
Hey @NielDuysters 

This PR was closed due to inactivity. If this change is still important to you, feel free to create a new pull request. 

For more information about our contribution guidelines, see https://github.com/shopware/shopware/blob/trunk/CONTRIBUTING.md

NielDuysters commented:
@mitelg Hi. Sorry for my inactivity. But coincidentally I was just given a full day at my dayjob to work on this PR. Could you consider reopening this? I am working on it as we speak.

mitelg commented:
@NielDuysters ah nice :grin: :+1: 

it is open again, happy to see the result :handshake:

NielDuysters committed:
entityindexerregistry: Add index-only flag to Sync API

NielDuysters committed:
Merge branch 'feat/entityindexerregistry-extension-indexer-only' of https://github.com/NielDuysters/shopware into feat/entityindexerregistry-extension-indexer-only

NielDuysters committed:
entityindexerregistry: Rename EntityIndexerRegistry::addOnlies to EntityIndexerRegistry::addOnlyAllowedIndices

NielDuysters committed:
entityindexerregistry: Update changelog with usage guidelines

NielDuysters commented:
@mitelg I am encountering issues with passing the CI check `"Danger / pr (pull_request_target)"`.
<img width="812" alt="afbeelding" src="https://github.com/user-attachments/assets/d6f83c2e-2044-4ed8-9f38-aecce7e6bf8d" />

However, running PHPStan locally does not yield any errors:
![afbeelding](https://github.com/user-attachments/assets/070f8feb-6685-4e66-9104-bbb560d0db45)

The only "error" my linter gives is:
<img width="713" alt="afbeelding" src="https://github.com/user-attachments/assets/449652cd-0bc8-492e-b50b-e63063d0ff76" />

`"Throwing new exceptions within classes are not allowed. Please use domain exception pattern"`, but this pattern is also not applied in other files so IG that's also not the reason the CI check fails.

mitelg commented:
this jobs checks if there are PHPStan issues in the baseline in files that were touched within a PR. normally we should try to fix those issues, but that is not necessary for externals

NielDuysters committed:
entityindexerregistry: Update test for SyncControllerTest

NielDuysters commented:
@mitelg I'm CIAD. WDYT?

mitelg commented on src/Core/Framework/Api/Sync/SyncBehavior.php:
```diff
@@ -9,10 +9,12 @@ class SyncBehavior
         protected ?string $indexingBehavior = null,
-        protected array $skipIndexers = []
+        protected array $skipIndexers = [],
+        protected array $skipOnlies = []
```
I think this name here is a bit misleading :thinking: maybe `onlyIndexers`? the getter should also be renamed then

mitelg commented on tests/integration/Core/Framework/Api/Controller/SyncControllerTest.php:
```diff
@@ -488,6 +488,41 @@ public function testSkipIndexer(): void
+        $connection = static::getContainer()->get(Connection::class);
+
+        $count = (int) $connection->fetchOne('SELECT COUNT(*) FROM product_search_keyword WHERE product_id = ?', [Uuid::fromHexToBytes($id1)]);
+        static::assertGreaterThan(0, $count, 'Search keywords should not empty as we did not skip it');
```
can we somehow assert, that other indexers did not run?

NielDuysters replied:
@mitelg What do you advice on how to write this test? Would it be okay to check if only a few indexers did not run and use that assert to derive that the other indexers also didn't run? Or is it really necessary to test ALL indexes?

mitelg replied:
I wonder if there is a way to check which indexers were executed. without querying the database for the tables they usually affect. something like checking the dispatched events or messages

mitelg replied:
I just checked again: would it be possible to register a listener on the `ProgressFinishedEvent` and collect the executed indexer names? after the run, there should only be the search keyword indexer in it.

NielDuysters replied:
I’ll check this. Thanks for the suggestion.

mitelg replied:
hey @NielDuysters 

did you already had the chance to have a look?

mitelg commented:
hey @NielDuysters 

what is the status here? :blush:

NielDuysters commented:
> hey @NielDuysters
> 
> what is the status here? 😊

@mitelg I will have time to work in this again friday. But I am still wondering how to solve the testing-scenario's to verify no other indexers did run.

mitelg commented on tests/integration/Core/Framework/Api/Controller/SyncControllerTest.php:
```diff
@@ -488,6 +488,41 @@ public function testSkipIndexer(): void
+        $connection = static::getContainer()->get(Connection::class);
+
+        $count = (int) $connection->fetchOne('SELECT COUNT(*) FROM product_search_keyword WHERE product_id = ?', [Uuid::fromHexToBytes($id1)]);
+        static::assertGreaterThan(0, $count, 'Search keywords should not empty as we did not skip it');
```
```suggestion
        static::assertGreaterThan(0, $count, 'Search keywords should not empty as it was the only called indexer');
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/shopware/shopware?pullRequest=8929) <br/>All committers have signed the CLA.

NielDuysters committed:
Update tests/integration/Core/Framework/Api/Controller/SyncControllerTest.php

Co-authored-by: Michael Telgmann <mitelg@users.noreply.github.com>

NielDuysters commented:
Applied following suggestion: https://github.com/shopware/shopware/pull/8929/commits/f8eb050e50816258d9ed68695c7183074928279f

mitelg committed:
Merge branch 'trunk' into feat/entityindexerregistry-extension-indexer-only

mitelg committed:
smaller code adjustments

mitelg commented:
hey @NielDuysters 

to get that PR finally done, I started to fix the open things and tried to find a possible way to test the behaviour properly. I have not finished this yet, but I already discovered, that your addition probably still does not work like intended. 

I added 
```
        print_r($name);
        echo "\n";
```
at the beginning of `\Shopware\Core\Profiling\Profiler::trace` and saw that still those indexers are executed:
```
product:indexer:inheritance
product:indexer:stock
product:indexer:search-keywords
product:indexer:states
```

I executed the new test `\Shopware\Tests\Integration\Core\Framework\Api\Controller\SyncControllerTest::testOnlyIndexer`

so my assumption is that the `\Shopware\Core\Framework\DataAbstractionLayer\Indexing\EntityIndexingMessage::allow` does not work as expected :thinking: 

Maybe you can also have a look again here.

NielDuysters commented:
@mitelg Thanks for helping me with the testing, I was stuck on that.

Weird that `\Shopware\Core\Framework\DataAbstractionLayer\Indexing\EntityIndexingMessage::allow` does not work as expected, I'll take a look into that. Is their maybe a reason you can think of of why these specific indexers keep running while other's don't?

mitelg commented:
@NielDuysters sorry, I did not dig any further than that :see_no_evil: I have no clue

NielDuysters commented:
@mitelg I rapidly looked into this. I actually think the code is working as expected. Looking at `ProductIndexer::update` located in [ProductIndexer.php:L90](https://github.com/shopware/shopware/blob/trunk/src/Core/Content/Product/DataAbstractionLayer/ProductIndexer.php#L90) it seems there are some "special" indexers like e.g the `product:indexer:inheritance` and `product:indexer:stock` you mentioned which consistently execute.

> I added (...) at the beginning of (...)

I did the same to reproduce your findings and got the same result. But if in `ProductIndexer` I comment out L98-L106:
```
        /*
        Profiler::trace('product:indexer:inheritance', function () use ($ids, $event): void {
            $this->inheritanceUpdater->update(ProductDefinition::ENTITY_NAME, $ids, $event->getContext());
        });

        $stocks = $event->getPrimaryKeysWithPropertyChange(ProductDefinition::ENTITY_NAME, ['stock', 'isCloseout', 'minPurchase']);
        Profiler::trace('product:indexer:stock', function () use ($stocks, $event): void {
            $this->stockStorage->index(array_values($stocks), $event->getContext());
        });
         */
```

and run the test again, only `product:indexer:search-keywords` is printed. So it seems this code is responsible to execute specific (seemingly important) indexers in any case.

AFAIU Shopware has "special" indexers which it runs to keep data consistent and correct. I suppose there is a good reason for why e.g `product:indexer:inheritance` and `product:indexer:stock` always run and I don't think we need to skip these indexers even when using the `EntityIndexerRegistry::EXTENSION_INDEXER_ONLY` from this PR to avoid breaking the internal workings of indexing.

Using `EntityIndexerRegistry::EXTENSION_INDEXER_ONLY` is supposed to allow better performance when doing updates (which it currently does) - but I don't think we need to take `INDEXER_ONLY` literally and also skip these few "important indexers" to avoid altering the mechanism of Shopware's indexing too much. Skipping these "important indexers" might result in unexpected behavior or inconsistent data.

mitelg commented:
hey @NielDuysters 

I talked with @shyim as I was also not sure about it, but yes you are right. those two are always called directly and the other indexer run async. 

the question now is, if we are able to test the properly within an integration test :grimacing: 
Soner had the wild idea to define a custom profiler, as the the indexer calls are traced :thinking:

NielDuysters commented:
@mitelg  Wild idea indeed. ;p

Isn't it also possible to create a subscriber for the EntityIndexer only enabled in tests, and let the subscriber log what indexers were triggered?

mitelg commented:
yes, that sounds like a good idea :+1:

NielDuysters commented:
@mitelg Yesterday I investigated multiple options but none seemed to work. Also the `EntityIndexer` maybe is not the correct place to check this.

What I am currently thinking: Find a way to force the messages on the queue, and use a query like `SELECT * FROM messenger_messages` to see the contents of the messages and validate all the indexers are skipped except the required one.

I guess this is the closest we can. Otherwise we indeed do have to build a custom profiler and use that in the `ProductIndexer`, but these traces also just check if a specific indexer is in the allow or skip list...

mitelg commented:
@NielDuysters okay, let's try this

NielDuysters committed:
entityindexerregistry: Update test

NielDuysters commented:
@mitelg I made a new commit: https://github.com/shopware/shopware/pull/8929/commits/3e33803bbbe7492b4e9291cf52e42aa1ea14af27

But I encounter two issues:

1) `./vendor/bin/phpunit --testsuite integration --filter testSkipIndexer` returns
```
Failed asserting that 500 is identical to 200.
```

The error in the response is: `SQLSTATE[42S22]: Column not found: 1054 Unknown column \u0027language.active`. Why is this? Do I have to define header `'HTTP_' . PlatformRequest::HEADER_LANGUAGE_ID => Defaults::LANGUAGE_SYSTEM` in another way?

Do I have to use translations in my payload?

```
                        'translations' => [
                            Defaults::LANGUAGE_SYSTEM => [
                                'name' => 'bar',
                                'description' => 'foo',
                                'keywords' => 'fiz,baz',
                            ],
                        ],
```

Tried both, none worked.


2) `./vendor/bin/phpunit --testsuite integration --filter testOnlyIndexer` returns
```
Search keywords should not empty as it was the only called indexer
Failed asserting that '0' is greater than 0.
```

Maybe because we are now forcing it on the queue and the indexer is not actually executed when asserting? 

Some assistance required please.

mitelg committed:
Merge branch 'trunk' into feat/entityindexerregistry-extension-indexer-only

mitelg committed:
fix tests and PHPStan

mitelg committed:
Merge branch 'trunk' into feat/entityindexerregistry-extension-indexer-only

mitelg commented on src/Core/Framework/DataAbstractionLayer/DataAbstractionLayerException.php:
```diff
@@ -1097,6 +1098,13 @@ public static function unexpectedConstraintType(Constraint $constraint, string $
+            Response::HTTP_NOT_FOUND,
+            self::ENTITY_INDEXER_NOT_FOUND,
+            self::$couldNotFindMessage,
+            ['entity' => 'entity indexer', 'field' => 'name', 'value' => $name],
```
```suggestion
            ['entity' => 'entity indexer', 'field' => 'name', 'value' => $name],
         );
     }
```

mitelg committed:
Update src/Core/Framework/DataAbstractionLayer/DataAbstractionLayerException.php

mitelg committed:
fix cs

mitelg commented:
hey @NielDuysters 

I changed the test a bit. Would you mind having a look again? I removed the check, if the updater actually run. for me it is enough, that only one indexer would be executed.

mitelg committed:
move to new release info

mitelg committed:
Merge branch 'trunk' into feat/entityindexerregistry-extension-indexer-only

mitelg committed:
Merge branch 'trunk' into feat/entityindexerregistry-extension-indexer-only

NielDuysters commented:
@mitelg Thanks for your assistance.

The tests look good IMHO.

> if the updater actually run.

The issue in `testOnlyIndexer` is that the updater actually won't run because we do not consume the message. But it nicely asserts that the created message contains the content we expect, and I guess we can assume it will be consumed correctly. I guess there are other tests verifying the actual execution of indexers from a message, so we shouldn't assert that again in our new test?

mitelg commented:
> I guess there are other tests verifying the actual execution of indexers from a message, so we shouldn't assert that again in our new test?

that was exactly my thought :+1: 

So I will hit the merge button now, finally :smiley:

mitelg committed:
Merge branch 'trunk' into feat/entityindexerregistry-extension-indexer-only

NielDuysters commented:
@mitelg Cool! Thanks for your assistance and patience. This was a very educating experience for me.

mitelg committed:
Merge branch 'trunk' into feat/entityindexerregistry-extension-indexer-only

mitelg commented:
@NielDuysters no, thank you for your patience and willingness to go this journey! :see_no_evil: nice to hear that you learned something :blush: :+1: hope, we can see more contribution from you in the future :rocket: 

thanks again for your contribution! 👍🎉💙
